src/relax/transform/dataflow_inplace.cc

```diff
@@ -0,0 +1,1017 @@
+      ret->Remove(gv);
+    }
+    return ret;
+  }
+
+  // for handling inner functions
+  Expr VisitExpr_(const FunctionNode* op) override {
+    auto old_func_params = func_params;
+    func_params = op->params;
+    auto ret = ExprMutator::VisitExpr(GetRef<Function>(op));
+    func_params = old_func_params;
+    return ret;
+  }
+
+  // the only case we will override: we will visit all binding blocks
+  // and replace any valid calls in them
+  BindingBlock VisitBindingBlock_(const DataflowBlockNode* op) override {
+    auto block = GetRef<DataflowBlock>(op);
+    // std::unordered_set<Var, ObjectPtrHash, ObjectPtrEqual> free_var_set;
+    // for (auto binding : block->bindings) {
+    //   auto binding_free_vars = FreeVars(GetBoundValue(binding));
+    //   free_var_set.insert(binding_free_vars.begin(), binding_free_vars.end());
+    // }
+    // Array<Var> free_var_list(free_var_set.begin(), free_var_set.end());
+
+    // For now, only handle exact match cases.
+    // Note: Not passing any input values for now, as we can't make any assumptions
+    // about them.
+    auto matches_found = FindInplaceOpportunities(block, {});
+    auto exact_matches = matches_found.second;
+
+    Array<Binding> new_bindings;
+    int current_match_index = 0;
+    for (size_t i = 0; i < block->bindings.size(); i++) {
+      if (current_match_index >= static_cast<int>(exact_matches.size()) ||
+          exact_matches[current_match_index][0] != static_cast<int>(i)) {
+        new_bindings.push_back(block->bindings[i]);
+        continue;
+      }
+
+      auto target_binding = block->bindings[i];
+      auto target_call = Downcast<Call>(GetBoundValue(target_binding));
+      // can just pick the first index arbitrarily (only using one output for now too)
+      auto new_call = CreateInplaceCall(target_call, {exact_matches[current_match_index][1]});
+      // now replace the binding appropriately
+      if (auto* var_binding_node = target_binding.as<VarBindingNode>()) {
+        auto var_binding = GetRef<VarBinding>(var_binding_node);
+        auto* var_binding_cow = var_binding.CopyOnWrite();
+        var_binding_cow->value = new_call;
+        new_bindings.push_back(var_binding);
+      } else if (auto* match_cast_node = target_binding.as<MatchCastNode>()) {
+        auto match_cast = GetRef<MatchCast>(match_cast_node);
+        auto* match_cast_cow = match_cast.CopyOnWrite();
+        match_cast_cow->value = new_call;
+        new_bindings.push_back(match_cast);
+      } else {
+        CHECK(false) << "Invalid binding type";
+      }
+      current_match_index++;
+    }
+    return DataflowBlock(new_bindings, block->span);
+  }
+
+  // Given the call and indices of arguments that could be done in-place,
+  // replace the call with a call to an in-place PrimFunc.
+  // (Made public for testing.)
+  Call CreateInplaceCall(const Call& call, const Array<Integer>& inplace_indices) {
+    static const auto& legalize_map = Op::GetAttrMap<FLegalize>("FLegalize");
```

Comment:
Should this use a separate legalization map?  Even if they map to the same function, the per-operator support could then be checked by seeing if `"FLegalizeInPlace"` is defined for an operator.

Reply:
That's worth thinking about. I didn't use it here because the operators that can be done in-place can also be implemented by applying the transformation given here to their legalized version. @tqchen, @MasterJH5574 (and others), do you think it would be worthwhile to add another map for in-place legalizations?

Reply:
I would like for a pass like this to come later in compilation, but the issue is the fact it relies on working within single dataflow blocks, so it has to come before we get rid of dataflow blocks. That is also causing me a few issues with thinking about how to handle in-place split and concat.

Reply:
This is probably a larger change, but would it make sense to either (A) implement it in terms of bindings without requiring a dataflow block, or (B) implement the pass as `Sequential(ConvertToDataFlow(), current_pass(), ToNonDataflow())`?

(B) would be easier to implement, but the use of `ToNonDataflow` as a post-processor would require that it be after any passes that require dataflows to be present.

Reply:
If we don't want to have the dataflow requirement, we would also need to handle control flow. That's the main issue and there's no obvious way to deal with control flow without going with a solution more like abstract interpretation (see my comment about the general-purpose liveness analysis PR).

Reply:
Using the sequential transformation would be a less ad hoc version of the above.

Reply:
Good call, and agreed that the sequential transform would be the better way to handle it, and a more consistent pattern that could be applied to other passes.